bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/sds011protocol/SDS011Communicator.java

```diff
@@ -0,0 +1,297 @@
+        SensorReply reply = sendCommand(m);
+        logger.debug("Got reply to getFirmware command: {}", reply);
+
+        if (reply instanceof SensorFirmwareReply) {
+            SensorFirmwareReply fwReply = (SensorFirmwareReply) reply;
+            thingHandler.setFirmware(fwReply.getFirmware());
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Request data from the device, they will be returned via the serialEvent callback
+     *
+     * @throws IOException
+     */
+    public void requestSensorData() throws IOException {
+        CommandMessage m = new CommandMessage(Command.REQUEST_DATA, new byte[] {});
+        byte[] data = m.getBytes();
+        if (logger.isDebugEnabled()) {
+            logger.debug("Requesting sensor data, will send: {}", HexUtils.bytesToHex(data));
+        }
+        outputStream.write(data, 0, data.length);
+    }
+
+    private @Nullable SensorReply readReply() throws IOException {
+        byte[] readBuffer = new byte[Constants.REPLY_LENGTH];
+
+        int b = -1;
+        if (inputStream.available() > 0) {
+            while ((b = inputStream.read()) != Constants.MESSAGE_START_AS_INT) {
+                logger.debug("Trying to find first reply byte now...");
+            }
+            readBuffer[0] = (byte) b;
+            int remainingBytesRead = inputStream.read(readBuffer, 1, Constants.REPLY_LENGTH - 1);
+            if (logger.isDebugEnabled()) {
+                logger.debug("Read remaining bytes: {}, full reply={}", remainingBytesRead,
+                        HexUtils.bytesToHex(readBuffer));
+            }
+            return ReplyFactory.create(readBuffer);
+        }
+        return null;
+    }
+
+    /**
+     * Data from the device is arriving and will be parsed accordingly
+     */
+    @Override
+    public void serialEvent(SerialPortEvent event) {
+        if (event.getEventType() == SerialPortEvent.DATA_AVAILABLE) {
+            // we get here if data has been received
+            SensorReply reply = null;
+            try {
+                reply = readReply();
+                logger.debug("Got data from sensor: {}", reply);
+            } catch (IOException e) {
+                logger.warn("Could not read available data from the serial port: {}", e.getMessage());
+            }
+            if (reply instanceof SensorMeasuredDataReply) {
+                SensorMeasuredDataReply sensorData = (SensorMeasuredDataReply) reply;
+                if (sensorData.isValidData()) {
+                    thingHandler.updateChannels(sensorData);
+                }
+            }
+        }
+    }
+
+    /**
+     * Shutdown the communication, i.e. send the device to sleep and close the serial port
+     */
+    public void dispose() {
+        if (serialPort != null) {
+            try {
+                // send the device to sleep to preserve power and extend the lifetime of the sensor
+                sendSleep(true);
```

Comment:
shouldn't you remove the event listener before you call this?

Reply:
I tried this before, but somehow the underlieing library does something with the port if I de-register my listener and then my command doesn't go through.

This is just a guess from my observation. That is why I did not de-register the listener before that call here.

Also in the callback method for the listener I check for valid data, so it doesn't matter if the callback receives the reply to the sleep on shutdown.